text/0000-century-of-the-self-type.md

```diff
@@ -0,0 +1,178 @@
+trait Foo {
+    fn by_value(self: Self);
+    fn by_ref(self: &Self);
+    fn by_ref_mut(self: &mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_arc(self: Arc<Self>);
+}
+```
+
+## Recursive arbitrary receivers
+
+Like the rule for deref coercions, the rule for receivers is recursive. If type
+`T` implements `Deref` targeting type `U`, and type `U` implements `Deref`
+targeting `Self`, `T` is a valid receiver (and so on outward).
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_ref_to_rc(self: &Rc<Self>) { ... }
+}
+```
+
+## Object safety
+
+In order for these receivers to be object safe, some additional traits need to
+be implemented. Given a reference type `Ptr<dyn Trait>`, the compiler must be
+able to prove that `T: Unsize<dyn Trait>` implies `Ptr<T>:
+CoerceUnsized<Ptr<dyn Trait>>`. If the compiler can prove this, methods with
+these receivers are object safe (how they object safe conversion is implemented
+is discussed later in the detailed design).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Handling trait objects
+
+To support object-safety, we must have a way of obtaining a vtable from the
+reference type, and then passing the correct receiver type to the function in 
+the vtable.
+
+### Step 1: Obtaining the vtable
+
+First, we call `Deref::deref` on the receiver type to obtain an `&Trait`
+object. This will return an `&dyn Trait`. Since an `&dyn Trait` is defined as 2
+words, one a pointer to the object data and the other a pointer to the vtable,
+we can obtain the vtable by dereferencing the second pointer of this object.
+
+For example, consider this type:
+
+```rust
+struct Foo<T: ?Sized> {
+    inner: T,
+}
+
+impl<T: ?Sized> Deref for Foo<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.inner
+    }
+}
+
+trait Bar {
+    fn baz(self: Foo<Self>);
+}
+```
+
+Here, we start with an `&Foo<Trait>`, which is represented as a wide pointer.
+When we call `Deref::deref`, we receive a wide pointer, with the data pointer
+pointing to `self.inner`, and the vtable pointing to that same vtable.
+
+### Step 2: Obtaining the correct receiver type
+
+Having obtained the vtable, we now need to obtain a value of the correct type
+to pass to the function. For example, given a trait like this:
+
+```rust
+trait Foo {
+    fn bar(self: Rc<Self>);
+}
+```
+
+The function in the vtable expects a value of `Rc<Self>`, where `Self` is the
+concrete type that was cast into the vtable. So if we have an `Rc<Foo>`, we
+need to temporarily cast that *back* into an `Rc<i32>` or whatever the concrete
+type is.
+
+This is why the `CoerceUnsized` bound is necessary for object-safe receiver
+types. Using the type ID stored in the vtable, we can downcast the `Self` type
+to the correct concrete type to pass to the function pointer for the method
+we're calling, effectively reversing the unsizing coercion.
```

Comment:
Are you sure that "type ID" and "downcast" is the right terminology here? My understanding was that the compiler looks up the method in the vtable, coerces the `Rc<Foo>` to `Rc<WhateverTypeThatImplementsFoo>` without actually knowing what that type is, and calls the method using the coerced `Rc` as the `self` argument

Reply:
You're right.